### DISCUSSION TITLE
How to navigate DateTime?

### DISCUSSION THREAD

**Discussion by JunoNgx** [pocketbase/pocketbase#4822] [Q&A] [answered]
Context: I'm performing a daily `cron` job to check if a soft-deleted item has been deleted for more than 7 days. If yes, then proceed to delete it permanently. Also, I'm using JS hooks.

In the custom endpoint, I'm managed to use `new DateTime()` to the time of soft-deletion in the column `trashedDateTime`.

I'm attempting to get the "7 days ago" timestamp with (while iterating through all the record in the db):

```
const sevenDaysAgoTime = new DateTime.time().addDate(0, 0, -7);
```

But I've been running into an error message `TypeError: Value is not an object: undefined`. Not sure if I'm missing anything, but I'm seeing from [the documentation](https://pocketbase.io/jsvm/classes/DateTime.html#time) that `DateTime.time()` should give me a `Time` instance.

Is my approach sane? Am I missing anything? Is there a more sane way to navigate datetime comaprison?

**Comment by ganigeorgiev** [comment]
You are missing parenthesis for the `DateTime` constructor, aka. it must be `new DateTime()`.

I'm not sure that I understand what you are trying to do but an alternative to the Go `DateTime` instance could be to operate with plain JS `Date`, aka. `const d = new Date()`.

If you need further help, please share a more detailed code sample to understand what _"while iterating through all the record in the db"_ means.

**Reply by JunoNgx** [reply]
Thank you, the correct mistake was not calling the `DateTime` constructor correctly with the parentheses.

Let me share the entire function that is passed into the cron job here:

```
handleExpiredTrashItems() {
    console.log("handleExpiredTrashItems")
    $app.logger().info(
        "Performing daily check of trashed items",
        "datetime", new Date().toISOString()
    );

    const allItems = $app.dao().findRecordsByExpr(types.DbTables.ITEMS);

    for (const item of allItems) {
        if (!item.getBool("isTrashed")) {
            continue;
        }

        try {
            const sevenDaysAgoTime = new DateTime().time().addDate(0, 0, -7);
            const trashedTime = item.getTime("trashedDateTime"); // isZero somehow
            if (trashedTime.compare(sevenDaysAgoTime) === 0
                || trashedTime.compare(sevenDaysAgoTime) === -1
            ) {
                $app.dao().deleteRecord(item);
            }
        } catch (err) {
            $app.logger().error(
                "Error attempting to delete trashed item",
                "itemId", item.id,
                "error", err.toString(),
            );
            console.log(err.toString());
        }

    }
}
```

Any idea why is `item.getTime("trashedDateTime")` returns `0001-01-01 00:00:00 +0000 UTC -1`, which I guess is a `isZero()` value? I'm pretty sure the these were set correctly in the DB (when the items were soft-deleted), just not sure what went wrong.

![Screenshot from 2024-04-24 20-04-25](https://github.com/pocketbase/pocketbase/assets/5671813/0d945d48-df32-426b-8ad6-20713d27ecf4)

Also, I guess there is some db expression that would help me get all the `isTrashed` items without me having to check them in the loop? These look very foreign to me, so I'm not comfortable with them at the moment.

**Reply by ganigeorgiev** [reply, marked as answer]
`item.getTime("trashedDateTime")` returns a zero time most likely because it is failing to cast the string value into a time.Time (_this method may get removed with the refactoring to avoid the confusion with `getDateTime`_). An alternative could be to get a `DateTime` because that's actually the underlying value, aka. `item.getDateTime("trashedDateTime")`.

In any case if your requirement is to fetch records that are `isTrashed=true` and `trashedDateTime` older than 7 days you could try:

```js
/// <reference path="../pb_data/types.d.ts" />

function handleExpiredTrashItems() {
    console.log("handleExpiredTrashItems")
    $app.logger().info(
        "Performing daily check of trashed items",
        "datetime", new Date().toISOString()
    );

    const sevenDaysAgo = new Date();
    sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 7); // 7 days from now, with the current time

    const trashedItems = arrayOf(new Record)

    // an alternative to the below query could be to use the filter expression with findRecordsByFilter
    $app.dao().recordQuery(types.DbTables.ITEMS)
        .andWhere($dbx.exp("isTrashed = true AND trashedDateTime < {:date}", {
            date: sevenDaysAgo.toISOString().replace("T", " "), // replace is needed to convert the date to RFC3339
        }))
        .all(trashedItems);

    for (const item of trashedItems) {
        try {
            // before enabling the deleteRecord call verify first that
            // the above query returns the items you expect
            console.log(item.get("trashedDateTime"))
            // $app.dao().deleteRecord(item);
        } catch (err) {
            $app.logger().error(
                "Error attempting to delete trashed item",
                "itemId", item.id,
                "error", err.toString(),
            );
        }
    }
}
```

**Reply by JunoNgx** [reply]
I got an error from your snippet `GoError: SQL logic error: near "&": syntax error (1); failed query: SELECT `items`.* FROM `items` WHERE isTrashed = true && trashedDateTime < {:date}`, though unfortunately I'm not well versed in this enough to tell what went wrong.

**Reply by JunoNgx** [reply]
The alternative you offered, using `item.getDateTime("trashedDateTime")`, unfortunately doesn't have the `compare` method I'm using to compare against `sevenDaysAgoTime`, so I'm getting the error `TypeError: Object has no member 'compare'`.

Is there a way to compare `Time` with `DateTime` or `DateTime` with `DateTime`?

**Reply by ganigeorgiev** [reply]
Sorry, replace `&&` with `AND`.

**Reply by JunoNgx** [reply]
Looks like getting `const trashedTime = item.getDateTime("trashedDateTime").time()` works, but that's still pretty mouthful, and I personally don't feel like there is a comfortable interface to work with dateTime yet.

Nevertheless, thank you very much for your assistance. I have managed to implement what I needed to. I hope this is taken into account as feedback for future development.
